Network/Socket/Internal.hsc

```diff
@@ -271,3 +280,26 @@ withSocketsInit = unsafePerformIO $ do
 foreign import ccall unsafe "initWinSock" initWinSock :: IO Int
 
 #endif
+
+wrapCheckStatus :: (Socket -> IO a) -> String -> Socket -> IO a
+wrapCheckStatus act fnName sock@(MkSocket _ _ _ _ statusVar) =
+  withMVar statusVar $ \status ->
+    case status of
+      Closed -> throwIO $ userError $
```

Comment:
I'm not the biggest fan of punning on `UserError`, but I think that ship has sailed, this convention has already been established.